lib/go-tc/invalidationjobs.go

```diff
@@ -456,3 +463,210 @@ func (job *UserInvalidationJobInput) Validate(tx *sql.Tx) error {
+		ID,
+		AssetURL,
+		CreatedBy,
+		DeliveryService,
+		TTLHours,
+		InvalidationType,
+		StartTime)
+}
+
+// Validate checks that the InvalidationJob is valid, by ensuring all of its fields are well-defined.
+//
+// This returns an error describing any and all problematic fields encountered during validation.
+func (job *InvalidationJobV40) Validate() error {
+	errs := []string{}
+	err := validation.ValidateStruct(job,
+		validation.Field(&job.DeliveryService, validation.Required),
+		validation.Field(&job.AssetURL, validation.Required, is.URL),
+		validation.Field(&job.CreatedBy, validation.Required),
+		validation.Field(&job.ID, validation.Required),
+		validation.Field(&job.TTLHours, validation.Required),
+		validation.Field(&job.InvalidationType, validation.Required, validation.NewStringRule(func(s string) bool {
+			return s == REFRESH || s == REFETCH
+		}, fmt.Sprintf("must be either %s or %s (case sensitive)", REFRESH, REFETCH))),
+	)
+
+	if err != nil {
+		errs = append(errs, err.Error())
+	}
+
+	if job.StartTime == nil {
+		return errors.New(strings.Join(append(errs, "startTime: cannot be blank"), ", "))
+	}
+
+	if job.StartTime.After(time.Now().Add(twoDays)) {
+		errs = append(errs, "startTime: must be within two days from now")
+	}
+
+	if job.StartTime.Before(time.Now()) {
+		errs = append(errs, "startTime: cannot be in the past")
+	}
+
+	if len(errs) > 0 {
+		return errors.New(strings.Join(errs, ", "))
+	}
+
+	return nil
+}
+
+// validateDeliveryService ensures the supplied (required) Delivery Service XML ID exists
+func (job *InvalidationJobCreateV40) validateDeliveryService(tx *sql.Tx) error {
+	var exists bool
+	row := tx.QueryRow(`SELECT EXISTS(SELECT * FROM deliveryservice WHERE xml_id=$1)`, job.DeliveryService)
+	if err := row.Scan(&exists); err != nil {
+		if err == sql.ErrNoRows {
+			return fmt.Errorf("No DeliveryService exists matching identifier: %v", job.DeliveryService)
+		}
+		return errors.New("Unknown error occurred")
+	}
+	return nil
+}
+
+// validateTLLHours ensures the supplied TTL hours is within acceptable limits
+func (job *InvalidationJobCreateV40) validateTLLHours(tx *sql.Tx) error {
+	var maxDays uint
+	err := tx.QueryRow(`SELECT value FROM parameter WHERE name='maxRevalDurationDays' AND config_file='regex_revalidate.config'`).Scan(&maxDays)
+	maxHours := maxDays * 24
+	if err == nil && uint(job.TTLHours) > maxHours { // silently ignore other errors too
+		return fmt.Errorf("cannot exceed %s", strconv.FormatUint(uint64(maxHours), 10))
+	}
+	return nil
+}
+
+func (job *InvalidationJobCreateV40) GetDSIDfromDSXMLID(tx *sql.Tx) (uint, error) {
+	var dsID uint
+	row := tx.QueryRow(`SELECT id FROM deliveryservice WHERE xml_id=$1`, job.DeliveryService)
+	if err := row.Scan(&dsID); err != nil {
+		if err == sql.ErrNoRows {
+			return 0, fmt.Errorf("No DeliveryService exists matching identifier: %v", job.DeliveryService)
+		}
+		return 0, errors.New("Unknown error occurred")
```

Comment:
well it wasn't an unknown error, we just dropped it. This should return that error. I think the reason you're not doing that is because you don't want to expose database errors to the client, and you're returning an error that tells that the DS doesn't exist which will be shown to the client. This is why `dbhelpers` functions typically return _three_ values for things like this - the requested value, whether or not it could be found, and whatever database error might have occurred, which is assumed to be unsafe for the client to view.